### ISSUE TITLE
App crashes when updating trip - NullPointerException in TripDetailActivity.isTheSameTrip()

### ISSUE THREAD

**Issue by ecanja** [grote/Transportr#316]
Version: 1.1.7

When updating a connection / a trip in the TripDetailActivity, the app crashes with the error message "Unfortunately, Transportr has stopped". `adb logcat` gives the following stacktrace:

```
E/AndroidRuntime(27760): FATAL EXCEPTION: main
E/AndroidRuntime(27760): Process: de.grobox.liberario, PID: 27760
E/AndroidRuntime(27760): java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Integer.equals(java.lang.Object)' on a null object reference
E/AndroidRuntime(27760): 	at de.grobox.liberario.activities.TripDetailActivity.isTheSameTrip(TripDetailActivity.java:221)
E/AndroidRuntime(27760): 	at de.grobox.liberario.activities.TripDetailActivity.onTripRetrieved(TripDetailActivity.java:152)
E/AndroidRuntime(27760): 	at de.grobox.liberario.tasks.AsyncQueryTripsTask.onPostExecute(AsyncQueryTripsTask.java:152)
E/AndroidRuntime(27760): 	at de.grobox.liberario.tasks.AsyncQueryTripsTask.onPostExecute(AsyncQueryTripsTask.java:42)
E/AndroidRuntime(27760): 	at android.os.AsyncTask.finish(AsyncTask.java:636)
E/AndroidRuntime(27760): 	at android.os.AsyncTask.access$500(AsyncTask.java:177)
E/AndroidRuntime(27760): 	at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:653)
E/AndroidRuntime(27760): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(27760): 	at android.os.Looper.loop(Looper.java:135)
...
```

As far as I can remember, this error has been introduced in 1.1.7 and has not occurred before. I hope that the stracktrace helps fixing it.

**Comment by grote** [comment]
Thanks for reporting this!

It is strange, because this code shouldn't have changed for 1.1.7. Can you reliably reproduce this crash all the time when refreshing a trip?

**Comment by ecanja** [comment]
Well, kind of. It's reliably happening all the time with the BVG network in Berlin.
After switching to another network (DB), everything's fine and no crash is happening.

**Comment by ecanja** [comment]
After having read #190 I realized that XPosed might be the culprit and more information might be useful for you:

1. I experience this problem on two devices.
2. A few weeks ago, I updated the OS of both devices (and installed microG and an XPosed module).
3. During this update, device 1 also got XPosed for the first time. I'm sure that the "updating trip"-feature had been fine on this device months ago (but I can't narrow down the occurrence of the exception to the installation of XPosed, the update of Transportr or any other event).
(Points 2 and 3 represent the only major difference on the phones that I can remember compared to weeks or months ago).
4. Device 2 has had XPosed installed for many months now and I think that I had been using the feature without problems there as well, but in fact I'm not sure about this anymore - it might be possible that it has never worked with XPosed (or at least with the newest XPosed version).
5. The whole stacktrace:

```E/AndroidRuntime(18589): FATAL EXCEPTION: main
E/AndroidRuntime(18589): Process: de.grobox.liberario, PID: 18589
E/AndroidRuntime(18589): java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Integer.equals(java.lang.Object)' on a null object reference
E/AndroidRuntime(18589): 	at de.grobox.liberario.activities.TripDetailActivity.isTheSameTrip(TripDetailActivity.java:221)
E/AndroidRuntime(18589): 	at de.grobox.liberario.activities.TripDetailActivity.onTripRetrieved(TripDetailActivity.java:152)
E/AndroidRuntime(18589): 	at de.grobox.liberario.tasks.AsyncQueryTripsTask.onPostExecute(AsyncQueryTripsTask.java:152)
E/AndroidRuntime(18589): 	at de.grobox.liberario.tasks.AsyncQueryTripsTask.onPostExecute(AsyncQueryTripsTask.java:42)
E/AndroidRuntime(18589): 	at android.os.AsyncTask.finish(AsyncTask.java:636)
E/AndroidRuntime(18589): 	at android.os.AsyncTask.access$500(AsyncTask.java:177)
E/AndroidRuntime(18589): 	at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:653)
E/AndroidRuntime(18589): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(18589): 	at android.os.Looper.loop(Looper.java:135)
E/AndroidRuntime(18589): 	at android.app.ActivityThread.main(ActivityThread.java:5254)
E/AndroidRuntime(18589): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(18589): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime(18589): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
E/AndroidRuntime(18589): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
E/AndroidRuntime(18589): 	at de.robv.android.xposed.XposedBridge.main(XposedBridge.java:102)```

Does this rather look like an XPosed issue from your perspective?

PS: I really appreciate this open-source public transport app and I'm impressed by the effort you're putting into it. Great work! I'm really glad that this project exists!

**Comment by grote** [comment]
At least the stacktrace looks like it originates from Xposed. Is there any way you can remove/disable it and then try without?

**Comment by ecanja** [comment]
Ha, you're right. I didn't think about the possibility to disable XPosed and just try it way quicker that way (I wanted to wait for the next OS update which is due this week ...). Did it now, but unfortunately it had no effect and the exception still pops up with the same stracktrace except for the last line. I'm nevertheless going to try it without XPosed after the next OS update (as far as I have time for this ...).

btw, all stacktraces I've seen so far on my phone do include XPosed in (one of) the last line(s), though in (almost) all cases I've seen so far XPosed is not responsible for the exception. It's just included because XPosed is always present - namely XPrivacy jumps into every process and is always sitting in between the app's process and the system...

However, if XPosed is not the culprit, the error should be reproducable (by others / you). If it's not, what else could influence this behaviour?

**Comment by patkan** [comment]
I am pretty sure that this is not related to Xposed. If I follow the description by @ecanja I can reproduce the bug and I have Android 5.1 that is rather close to AOSP. (Fairphone FP2 with stock rom).

If I find the time I will try to step through the software and see if I can find something.

**Comment by grote** [comment]
Might be due to the change of the BvgProvider to a different API recently.

**Comment by ecanja** [comment]
As @patkan says, it's not related to XPosed (I've updated my system yesterday and tried without XPosed - same behaviour).

@grote In this case, should we pass this bug report to [PTE](https://github.com/schildbach/public-transport-enabler) then?

**Comment by patkan** [comment]
The problem occurs in `TripDetailActivity.java:221` for seemingly all provider that are based on `AbstractHafasMobileProvider`. (I checked VMT and BVG, which are based on `AbstractHafasMobileProvider`, the problem does not occur in `AbstractHafasLegacyProvider`. I checked DB and NASA for this and could not witness the crash)

The problem is in this line:
```
if(old_trip.numChanges.equals(new_trip.numChanges) && old_trip.legs.size() == new_trip.legs.size()) {
```
In the case of `AbstractHafasMobileProvider` `numChanges` is `null`. In other cases it contains integer-values. So I would say we have to check whether `numChanges` may be `null` (I could not find `@Nullable`-Annotators.) and if yes we have to adapt our code. If no, then it is a bug in PTE.

If I have a look at `de/schildbach/pte/AbstractHafasMobileProvider.java:547`
```
final Trip trip = new Trip(null, tripFrom, tripTo, legs, fares, null, null);
```
, which is the line where the trips are added to the output in `AbstractHafasMobileProvider`, then this seems intentional (the last parameter would be the number of changes). Maybe the new API does not provide the value.
